core/tests/puppeteer-acceptance-tests/utilities/user/logged-out-user.ts

```diff
@@ -2846,6 +2865,241 @@ export class LoggedOutUser extends BaseUser {
+  }
+
+  /**
+   * Checks if the progress URL validity info matches the expected text. (To be used when save progress modal is opened.)
+   * @param {string} expectedText - The expected validity info text.
+   */
+  async checkProgressUrlValidityInfo(expectedText: string): Promise<void> {
+    const validityInfoText = await this.page.evaluate(selector => {
+      const element = document.querySelector(selector);
+      return element ? element.textContent.trim() : null;
+    }, validityInfoTextSelector);
+
+    if (validityInfoText !== expectedText) {
+      throw new Error(
+        `Validity info text does not match expected text. Found: ${validityInfoText}, Expected: ${expectedText}`
+      );
+    }
+  }
+
+  /**
+   * Copies the progress URL to the clipboard and returns the copied text. (To be used when save progress modal is opened.)
+   */
+  async copyProgressUrl(): Promise<string> {
+    try {
+      // OverridePermissions is used to allow clipboard access.
+      const context = await this.page.browser().defaultBrowserContext();
+      await context.overridePermissions('http://localhost:8181', [
+        'clipboard-read',
+        'clipboard-write',
+      ]);
+
+      // Click on the copy button.
+      await this.clickOn(copyProgressUrlButton);
+
+      // Reading the clipboard data.
+      const clipboardData = await this.page.evaluate(async () => {
+        return await navigator.clipboard.readText();
+      });
+
+      return clipboardData;
+    } catch (error) {
+      console.error('An error occurred:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Starts an exploration with a progress URL.
+   * @param {string} progressUrl - The URL to navigate to.
+   */
+  async startExplorationUsingProgressUrl(progressUrl: string): Promise<void> {
+    await this.goto(progressUrl);
+  }
+
+  /**
+   * Checks if the lesson info shows the expected rating.
+   * @param {string} expectedRating - The expected rating.
+   */
+  async expectLessonInfoToShowRating(expectedRating: string): Promise<void> {
+    const ratingText = await this.page.evaluate(selector => {
+      const element = document.querySelector(selector);
+      return element ? element.textContent.trim() : null;
+    }, ratingContainerSelector);
+
+    if (ratingText !== expectedRating) {
+      throw new Error(
+        `Rating text does not match expected rating. Found: ${ratingText}, Expected: ${expectedRating}`
+      );
+    }
+  }
+
+  /**
+   * Checks if the lesson info shows the given contributor.
+   * @param {string} contributorName - The name of the contributor.
+   */
+  async expectLessonInfoToShowContributors(
+    contributorName: string
+  ): Promise<void> {
+    const contributors = await this.page.$$eval(
+      `${contributorsContainerSelector} li`,
+      (liElements, contributorName) => {
+        return liElements.some(li => {
+          const profileLinkImage = li.querySelector('profile-link-image');
```

Comment:
Can this and the one below set as constant?

Reply:
Done.